### PR TITLE
Blackout/syn 5252/passwd policy epiphyte

### PR DESCRIPTION
**Pull request by vEpiphyte**
- password policy docs
- allow disabling policy groups
- additional tests with some more example policies

### CODE REVIEW

File: docs/synapse/devopsguide.rst [hunk 1]
```diff
@@ -499,6 +499,110 @@ Once they are enrolled, the user can connect using the Telepath URL ``aha://cort
...
+    Maximum number of incorrect attempts before locking user account. This is an integer value.
+
+``previous``
+    Number of previous passwords to disallow. This is an integer value.
+
+``complexity``
+    The complexity key must be set to an object. It can have the following keys:
+
+    ``length``
+        Minimum password character length. This is an integer value.
```
Comment (MichaelSquires) [inline docs/synapse/devopsguide.rst:522]:
```suggestion
        Minimum password character length. This is an integer value with a minimum value of 1.
```

File: docs/synapse/devopsguide.rst [hunk 2]
```diff
@@ -499,6 +499,110 @@ Once they are enrolled, the user can connect using the Telepath URL ``aha://cort
...
+
+    ``sequences``
+        Maximum sequence length in a password. Sequences can be letters or number, forward or reverse. This is an
+        integer value.
```
Comment (MichaelSquires) [inline docs/synapse/devopsguide.rst:526]:
```suggestion
        integer value with a minimum value of 2.
```

File: docs/synapse/devopsguide.rst [hunk 3]
```diff
@@ -499,6 +499,110 @@ Once they are enrolled, the user can connect using the Telepath URL ``aha://cort
...
+
+    ``upper:count``
+        The minimum number of uppercase characters required in password. This is a string value.
```
Comment (MichaelSquires) [inline docs/synapse/devopsguide.rst:529]:
```suggestion
        The minimum number of uppercase characters required in password. This is an integer value.
```

File: docs/synapse/devopsguide.rst [hunk 4]
```diff
@@ -499,6 +499,110 @@ Once they are enrolled, the user can connect using the Telepath URL ``aha://cort
...
+
+    ``upper:valid``
+        All valid uppercase characters. This defaults to a string of uppercase ASCII characters. This can be set to
+        a null value to disable any checking of the uppercase characters rules.
+
+    ``lower:count``
+        The minimum number of lowercase characters required in password.
```
Comment (MichaelSquires) [inline docs/synapse/devopsguide.rst:536]:
```suggestion
        The minimum number of lowercase characters required in password. This is an integer value.
```

File: docs/synapse/devopsguide.rst [hunk 5]
```diff
@@ -499,6 +499,110 @@ Once they are enrolled, the user can connect using the Telepath URL ``aha://cort
...
+
+    ``lower:valid``
+        All valid lowercase characters. This defaults to a string of lowercase ASCII characters. This can be set to
+        a null value to disable any checking of the lowercase character rules.
+
+    ``special:count``
+        The minimum number of special characters required in password.
```
Comment (MichaelSquires) [inline docs/synapse/devopsguide.rst:543]:
```suggestion
        The minimum number of special characters required in password. This is an integer value.
```

File: docs/synapse/devopsguide.rst [hunk 6]
```diff
@@ -499,6 +499,110 @@ Once they are enrolled, the user can connect using the Telepath URL ``aha://cort
...
+
+    ``special:valid``
+        All valid special characters. This defaults to a string of ASCII punctuation characters. This can be set to
+        a null value to disable any checking of the special characters rules.
+
+    ``number:count``
+        The minimum number of digit characters required in password.
```
Comment (MichaelSquires) [inline docs/synapse/devopsguide.rst:550]:
```suggestion
        The minimum number of digit characters required in password. This is an integer value.
```

File: docs/synapse/devopsguide.rst [hunk 7]
```diff
@@ -499,6 +499,110 @@ Once they are enrolled, the user can connect using the Telepath URL ``aha://cort
...
+        user: "999"
+        image: vertexproject/synapse-cortex:v2.x.x
+        network_mode: host
+        restart: unless-stopped
+        volumes:
+            - ./storage:/vertex/storage
+        environment:
+            SYN_CORTEX_AXON: aha://axon...
+            SYN_CORTEX_JSONSTOR: aha://jsonstor...
+            SYN_CORTEX_AUTH_PASSWD_POLICY: '{"complexity": {"length": 12, "sequences": 3, "upper:count": 2, "lower:count": 2, "lower:valid": "abcdefghijklmnopqrstuvwxyz\\u03b1\\u03b2\\u03b3", "special:count": 2, "number:count": 2}, "attempts": 3, "previous": 2}'
```
Comment (MichaelSquires) [inline docs/synapse/devopsguide.rst:593]: Do the unicode characters have to be \u encoded or can the just be pasted directly?

File: synapse/lib/auth.py [hunk 8]
```diff
@@ -124,6 +124,8 @@ async def __anit__(self, slab, dbname, pref='', nexsroot=None, seed=None, maxuse
             slab (s_lmdb.Slab): The slab to use for persistent storage for auth
             dbname (str): The name of the db to use in the slab
         '''
+        if policy:
+            s_schemas.reqValidPasswdPolicy(policy)
```
Comment (MichaelSquires) [inline synapse/lib/auth.py:128]: Doesn't this happen already at startup?

**Review by MichaelSquires — CHANGES_REQUESTED** [review]

File: synapse/lib/auth.py (hunk 8 above)
Comment (vEpiphyte) [reply synapse/lib/auth.py:128]: With moving the schema from the confdef we would not longer be checking it at config time.

File: docs/synapse/devopsguide.rst (hunk 7 above)
Comment (vEpiphyte) [reply docs/synapse/devopsguide.rst:593]: We can embed them directly. That was a default JSON serialization.

File: docs/synapse/devopsguide.rst [hunk 9]
```diff
@@ -499,6 +499,110 @@ Once they are enrolled, the user can connect using the Telepath URL ``aha://cort
 
     python -m synapse.tools.storm aha://cortex.<yournetwork>
 
+.. _ devops-task-users-password:
+
+Managing Password Policies
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Services can be configured with password policies. These can be used to define the complexity of a password, the number
+of allowed login attemps, and the number of previous passwords to check against. This is configured by setting the
```
Comment (Cisphyx) [inline docs/synapse/devopsguide.rst:508]:
```suggestion
of allowed login attempts, and the number of previous passwords to check against. This is configured by setting the
```

**Review by MichaelSquires — APPROVED** [review]

**Review by Cisphyx — APPROVED** [review]

**Pull request merged**